tests/test_smc.py

```diff
@@ -0,0 +1,213 @@
+
+        # assert sum of weights before equals sum of weights after
+        log_sum_weight_before_resampling = logsumexp(log_weights)
+        log_sum_weight_after_resampling = logsumexp(resampled_log_weights)
+        np.testing.assert_almost_equal(log_sum_weight_after_resampling, log_sum_weight_before_resampling)
+
+        # assert resampled_indices are all between 0 and n_particles
+        assert min(resampled_indices) >= 0
+        assert max(resampled_indices) < n_particles
+
+
+def test_effective_sample_size():
+    """On a collection of random log_weights vectors of varying size, assert that:
+    * ess >= 1
+    * ess <= n_particles
+
+    Also assert that:
+    * on a vector of constant log weights `zeros(n_particles) + constant`,
+        ess == n_particles, regardless of constant
+    * on a vector `[0, -inf, -inf, ...] + constant`,
+        ess == 1, regardless of constant
+    """
+    np.random.seed(2022)
+
+    # random weights: ESS should be between 1 and n_particles
+    n_instances = 100
+    for _ in range(n_instances):
+        n_particles = np.random.randint(1, 100)
+        log_weights = generate_log_weights(n_particles)
+        ess = effective_sample_size(log_weights)
+
+        assert ess >= 1
+        assert ess <= n_particles
+
+    # constant weights: ESS should be n_particles
+    for _ in range(n_instances):
+        n_particles = np.random.randint(1, 100)
+        constant = np.random.randn()
+        log_weights = np.zeros(n_particles) + constant
+        ess = effective_sample_size(log_weights)
+
+        np.testing.assert_almost_equal(ess, n_particles)
+
+    # degenerate weights: ESS should be 1
+    for _ in range(n_instances):
+        n_particles = np.random.randint(1, 100)
+        constant = np.random.randn()
+        index = np.random.randint(n_particles)
+        log_weights = np.log(np.zeros(n_particles))
+        log_weights[index] = constant
+
+        ess = effective_sample_size(log_weights)
+
+        np.testing.assert_almost_equal(ess, 1)
+
+
+def test_fractional_effective_sample_size():
+    """On random instances, check equal to effective_sample_size / n_particles"""
+    np.random.seed(2022)
+
+    # random weights
+    n_instances = 100
+    for _ in range(n_instances):
+        n_particles = np.random.randint(1, 100)
+        log_weights = generate_log_weights(n_particles)
+        ess = effective_sample_size(log_weights)
+        fractional_ess = fractional_effective_sample_size(log_weights)
+
+        np.testing.assert_almost_equal(fractional_ess, ess / n_particles)
+
+
+@pytest.mark.parametrize("resampling_fxn", [null_resample, multinomial_resample, conditional_multinomial_resample])
+def test_sequential_monte_carlo(resampling_fxn: Resampler):
+    """Run SMC with the desired resampling_fxn on a Gaussian 1D test system, and assert that
+    * running estimates of the free energy as a fxn of lambda match analytical free energies
+    * endstate samples have expected mean and stddev
```

Comment:
sentence fragment -- missing "and"?

Reply:
Replaced

```
assert that
* item1
* item2
* item3
```

with
```
assert that:
* item1,
* item2, and
* item3
```

in this file in https://github.com/proteneer/timemachine/pull/718/commits/af746deab1995f233fd247207eb046e7d23f5c3c (but may have missed other files)